cpp/unreal_plugins/UrdfBot/Source/UrdfParser.cpp

```diff
@@ -0,0 +1,381 @@
+            ASSERT(!robot_desc.joint_descs_.count(joint_desc.name_));
+            robot_desc.joint_descs_[joint_desc.name_] = std::move(joint_desc);
+        } else if (tag.Equals(TEXT("material"))) {
+            UrdfMaterialDesc material_desc = parseMaterialNode(child_node);
+            ASSERT(!robot_desc.material_descs_.count(material_desc.name_));
+            robot_desc.material_descs_[material_desc.name_] = std::move(material_desc);
+        }
+    }
+
+    // for a robot with N links, there must be N-1 joints.
+    ASSERT(robot_desc.joint_descs_.size() + 1 == robot_desc.link_descs_.size());
+
+    // build link tree based on joints
+    for (auto& joint_desc_pair : robot_desc.joint_descs_) {
+        UrdfJointDesc* joint_desc = &(joint_desc_pair.second);
+        ASSERT(joint_desc);
+
+        UrdfLinkDesc* parent_link_desc = &robot_desc.link_descs_.at(joint_desc->parent_link_name_);
+        UrdfLinkDesc* child_link_desc = &robot_desc.link_descs_.at(joint_desc->child_link_name_);
+
+        ASSERT(parent_link_desc);
+        ASSERT(child_link_desc);
+
+        // update the joint's pointers
+        ASSERT(!joint_desc->parent_link_desc_);
+        ASSERT(!joint_desc->parent_link_desc_);
+        joint_desc->parent_link_desc_ = parent_link_desc;
+        joint_desc->child_link_desc_ = child_link_desc;
+
+        // update the child's pointers (a child must have one parent)
+        ASSERT(!child_link_desc->parent_link_desc_);
+        ASSERT(!child_link_desc->parent_joint_desc_);
+        child_link_desc->parent_link_desc_ = parent_link_desc;
+        child_link_desc->parent_joint_desc_ = joint_desc;
+
+        // update the parent's pointers (a parent can have multiple children)
+        parent_link_desc->child_link_descs_.push_back(child_link_desc);
+        parent_link_desc->child_joint_descs_.push_back(joint_desc);
+    }
+
+    // for each link, update its material pointer, and find the root link
+    for (auto& link_desc_pair : robot_desc.link_descs_) {
+        UrdfLinkDesc* link_desc = &(link_desc_pair.second);
+        ASSERT(link_desc);
+
+        //bind material description
+        if (link_desc->material_name_ != "") {
+            UrdfMaterialDesc* material_desc = &robot_desc.material_descs_.at(link_desc->material_name_);
+            ASSERT(material_desc);
+
+            // update the link's material pointer (a link can have at most one material pointer)
+            ASSERT(!link_desc->material_desc_);
+            link_desc->material_desc_ = material_desc;
+        }
+        //find the root link
+        if (!link_desc->parent_link_desc_) {
+            ASSERT(!robot_desc.root_link_desc_);
+            robot_desc.root_link_desc_ = link_desc;
+        }
+    }
+
+    // a robot must have only one root link
+    ASSERT(robot_desc.root_link_desc_);
+
+    return robot_desc;
+}
+
+// parse urdf nodes
+UrdfLinkDesc UrdfParser::parseLinkNode(FXmlNode* link_node)
```

Comment:
A link can have multiple visual and collision nodes.

Reply:
fixed